staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go

```diff
@@ -0,0 +1,146 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package feature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_31 = version.MustParseSemantic("3.4.31")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_13 = version.MustParseSemantic("3.5.13")
+
+	// DefaultFeatureSupportChecker is a shared global etcd FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalculate feature support and caches it.
+	CheckClient(ctx context.Context, c client, feature storage.Feature) error
+}
+
+type defaultFeatureSupportChecker struct {
+	lock                       sync.Mutex
+	progressNotifySupported    *bool
+	progresNotifyEndpointCache map[string]bool
+}
+
+func newDefaultFeatureSupportChecker() *defaultFeatureSupportChecker {
+	return &defaultFeatureSupportChecker{
+		progresNotifyEndpointCache: make(map[string]bool),
+	}
+}
+
+// Supports can check the featue from anywhere without storage if it was cached before.
+func (f *defaultFeatureSupportChecker) Supports(feature storage.Feature) (bool, error) {
+	switch feature {
+	case storage.RequestWatchProgress:
+		f.lock.Lock()
+		defer f.lock.Unlock()
+
+		return ptr.Deref(f.progressNotifySupported, false), nil
+	default:
+		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+	}
+}
+
+// CheckClient accepts client and calculate the support per endpoint and caches it.
+// It will return at any point if error happens or one endpoint is not supported.
+func (f *defaultFeatureSupportChecker) CheckClient(ctx context.Context, c client, feature storage.Feature) error {
+	switch feature {
+	case storage.RequestWatchProgress:
+		return f.clientSupportsRequestWatchProgress(ctx, c)
+	default:
+		return fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+
+	}
+}
+
+func (f *defaultFeatureSupportChecker) clientSupportsRequestWatchProgress(ctx context.Context, c client) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	for _, ep := range c.Endpoints() {
+		supported, err := f.supportsProgressNotifyEndpointLocked(ctx, c, ep)
+		if err != nil {
+			return err
+		}
+		if !supported {
+			f.progressNotifySupported = ptr.To(false)
+			return nil
+		}
+	}
+	f.progressNotifySupported = ptr.To(true)
```

Comment:
```suggestion
	if len(c.Endpoints()) > 0 {
		f.progressNotifySupported = ptr.To(true)
	}
```
What if `c.Endpoints()` is empty?

Reply:
Ups, nice catch i'll add more tests.